selfscape_insight/features/ip_loc.py

```diff
@@ -44,19 +58,200 @@
+        image_base64 = base64.b64encode(buffer.getvalue()).decode('utf-8')
+
+        # Create HTML content for popup with the image
+        html = f'<img src="data:image/png;base64,{image_base64}">'
+        
+        # Update the ip_maps dictionary with IP address and its corresponding graph HTML
+        ip_maps[ip] = html
+        
+        # Clear the current plot to prepare for the next iteration
+        plt.close()
+        
+    return ip_maps
+
+# Function to handle click events on markers
+def on_map_click(event):
+    lat, lon = event.latlng
+    
+# Iterate through waypoints and add markers
+def create_html(df):
+    min_lon, max_lon = -45, -35
+    min_lat, max_lat = -25, -15
+    mymap = folium.Map(location=[0, 0],
+                zoom_start=2,
+                min_lat=min_lat,
+                max_lat=max_lat,
+                min_lon=min_lon,
+                max_lon=max_lon,
+                no_touch=True
+                )
+    
+    
+    
+    unique_ip = df['ip_address'].unique()
+    ip_time = generate_graph(df)
+    
+    for index, row in df.groupby('ip_address')[['latitude', 'longitude']].first().iterrows():
+        popup_text = index  # Access IP address directly
+        folium.Marker(location=[row['latitude'], row['longitude']], popup=ip_time[index]).add_to(mymap)
+
+    mymap.add_child(folium.ClickForMarker(popup=None))
+    mymap.get_root().html.add_child(folium.Element("""
+    <script>
+        document.getElementsByClassName('folium-map')[0].onclick = function(event){
+            var lat = event.latlng.lat;
+            var lon = event.latlng.lng;
+            var kernel = IPython.notebook.kernel;
+            kernel.execute('generate_graph([' + lat + ', ' + lon + '], df)');
+        }
+    </script>
+    """))
+
+    # Display the map
+    return mymap
+
+
+#account_activity_v2
+def run(pathname):
+
+    with open(pathname, 'r') as file:
+        data = json.load(file)
+    
+    account_activity_v2_value = data.get("account_activity_v2")
+    
+    df = pd.DataFrame(account_activity_v2_value)
+    
     print("Running the ip_loc feature module")
-    return "The ip_loc module did stuff!"
+    
+    # Assuming df is your DataFrame
+
+    edited_df = create_df(df)
+    folium_html = create_html(edited_df)
+    folium_html.save("folium_occurance.html")
+    return "Wrote folium_occurance.html"
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser(prog='ip_loc',
                                      description='A short description of what your code does')
-    parser.add_argument('-account_activity_v2', metavar='ACCOUNT_ACTIVITY_V2_CSV',
-                        help='path to account_activity_v2 csv file', required=True)
+    parser.add_argument('-account_activity_v2', metavar='ACCOUNT_ACTIVITY_V2',
+                        help='path to account_activity_v2 json file', required=True)
     args = parser.parse_args()
-    print(run(args.account_activity_v2))
+    run(args.account_activity_v2)
```

Comment:
You need to provide *some* way to tell the user what your code did... Carter's topics `return` is a good example.